zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+                OrderId = OrderId,
+            >;
+
+        type PalletId: Get<PalletId>;
+
+        /// The event type for this pallet.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// Weights generated by benchmarks.
+        type WeightInfo: WeightInfoZeitgeist;
+    }
+
+    /// The current storage version.
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+    pub(crate) type AssetOf<T> = Asset<MarketIdOf<T>>;
+    pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+    pub(crate) type BalanceOf<T> =
+        <<T as Config>::AssetManager as MultiCurrency<AccountIdOf<T>>>::Balance;
+    pub(crate) type OrderOf<T> = Order<AccountIdOf<T>, BalanceOf<T>, MarketIdOf<T>>;
+    pub(crate) type MarketIdOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
+    pub(crate) type MomentOf<T> = <<T as Config>::MarketCommons as MarketCommonsPalletApi>::Moment;
+    pub(crate) type MarketOf<T> =
+        Market<AccountIdOf<T>, BalanceOf<T>, BlockNumberFor<T>, MomentOf<T>, BaseAsset>;
+    pub(crate) type AmmTradeOf<T> = AmmTrade<BalanceOf<T>>;
+    pub(crate) type OrderTradeOf<T> = OrderbookTrade<AccountIdOf<T>, BalanceOf<T>>;
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T>
+    where
+        T: Config,
+    {
+        /// A trade was executed.
+        HybridRouterExecuted {
+            /// The type of transaction (Buy or Sell).
+            tx_type: TxType,
+            /// The account ID of the user performing the trade.
+            who: AccountIdOf<T>,
+            /// The ID of the market.
+            market_id: MarketIdOf<T>,
+            /// The maximum price limit for buying or the minimum price limit for selling.
+            price_limit: BalanceOf<T>,
+            /// The asset provided by the trader.
+            asset_in: AssetOf<T>,
+            /// The amount of the `asset_in` provided by the trader.
+            /// It includes swap and external fees.
+            /// It is an amount before fees are deducted.
+            amount_in: BalanceOf<T>,
+            /// The asset received by the trader.
+            asset_out: AssetOf<T>,
+            /// The aggregated amount of the `asset_out` already received
+            /// by the trader from AMM and orderbook.
+            /// It is an amount after fees are deducted.
+            amount_out: BalanceOf<T>,
+            /// The external fee amount paid in the base asset.
+            external_fee_amount: BalanceOf<T>,
+            /// The swap fee amount paid in the base asset.
+            swap_fee_amount: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The specified amount is zero.
+        AmountIsZero,
```

Comment:
Can't find a test using this error.

Reply:
ba390a211e78700de75a68006d00c516d5d66ffc